### ISSUE TITLE
Refactor the React-based Boxed Expression Component to be able to delete the GWT-based one

### ISSUE THREAD

**Issue by tiagobento** [apache/incubator-kie-issues#87]
The new, React-based Boxed Expression Editor needed some refactoring before we could aim for deleting the old, GWT-based Boxed Expression Editor.

---

Live prototype: https://tiagobento.github.io/prototypes/bee/
_(Last update: 2023-Apr-12, SHA: a6e3461e730fc2eee85f714843522fb9ed6903f0)_

Working branch: https://github.com/tiagobento/kie-tools/tree/bee-reisizing-perf

---

This refactor has the goal of making it easier to evolve the component, making room for new features.
Things to improve on this refactoring:

- State management
- Selection mechanism
- Resizing mechanism
- Context menu mechanism
- Performance improvements
- Integration with the DMN Editor
- Integration with the DMN Runner

## WRAP-UP

- [x] Make sure that we have green builds on https://github.com/tiagobento/kie-tools/pull/83
- [x] Fix `CI :: Code formatting / check`.
- [x] Merge `kiegroup/kie-tools#main` branch with `tiagobento/kie-tools#bee-reisizing-perf`.
- [x] Fix VS Code IT -> `KIE Editors Integration Test Suite - DMN Editor`
- [x] Console error:
  > Failed to decode downloaded font: http://localhost:3015/6f75249988833dd9f4297c9781ad35ff.ttf
      localhost/:1 OTS parsing error: invalid sfntVersion: 1702391919
- [x] Make sure keyboard shortcuts / navigation works well on Windows and Linux too.
- [x] CSS differences on the DMN Editor compared to the showcase app on `boxed-expression-component`.
- [x] [DMN Editor] React-based Boxed Expression Component Management
- - [x] Set the React-based Boxed Expression Component as the default choice
- - [x] Related to the above point, standalone now needs this fix --> https://github.com/kiegroup/kie-issues/issues/84
- - [x] Improved link back to DRG;
- - [x] Improved GWT and React layers API
- [x] [DMN Editor] Fix relationship with property link
- [x] DMN Elements [DMN Editor] IDs change when DMN File is persisted
- - [x] Decision Table Rules (IDs are not managed in React)
- - [x] Function
- - [x] Literal Expression, Context, Relation, Invocation & List
- [x] (Luiz) DMN Runner
- - [x] Switching between files should keep the Table/Form open, depending on what was open before.
- - [x] Keydown events on DMN Runner table should not be propagated to the rest of KIE Sandbox.
- - [x] Ensure DMN Runner Table UX is the same as in the Decision Table on DMN Editor. e.g. ESC shortcut, commit edits after enter, trigger select field on enter
- - [x] The BeeTable keyboard shortcuts are triggered while editing a DMN Runner Table cell;
- - [x] Changing input type should erase the input;
- - [x] DMN Runner outputs column should change its type after changing decision node type;
- - [x] DMN Runner table headers should display the correct data types. E.g. ('days and time duration')
- - [x] Differentiate between isActive and isEditing when navigating on BeeTable.
- [x] Autofilling nested expressions doesn't work in some scenarios

## KNOWN ISSUES

Issues that are currently identified on Boxed Expression Editor, and should be fixed. Most of them are bugs and do not prevent users from authoring expressions, though.

##### General:

- [x] Fix yard-related packages. They're not compiling after changes on BEE.
- [x] yard now correctly compiles and builds, but the integration seems not fully implemented. (Decision table rows data doesn't show, but that issue is present in `main` as well - no regressions)
- [x] Fix/delete unit tests.
- [x] Fix IT tests.

##### Resizing:

- [x] Java function type weird behavior when resizing
- [x] DecisionTable/Relation delete columns resizing bug.
- [x] Double-clicking the resizeHandler of a table header cell should change with to fit data.
- [x] Revisit `WidthMaths.ts` after removing `@patternfly/react-table`. There are 1px too much everywhere.
- [x] Resizing table columns when the container is already at minWidth.
- [x] Add resizer handles to all column headers.
- [x] Resizing a FEEL Function implementation only works after the second click on the resize handle, for any type of nested expression.
- [x] It shouldn't be possible to resize the label columns on PMML Functions. The same we do for JAVA Functions.

##### Keyboard:

- [x] Arrow-right on upper headers goes beyond the limit. Limit is correct using headerGroups.

##### Copy/cut/paste:

- [x] Allow pasting arbitrary data on cells that contain literal expressions.

##### Input:

- [x] Can't edit by start typing after pressing backspace on a cell, without pressing Enter.
- [x] Closing HitPolicy selector without pressing Esc (e.g. clicking outside) breaks keyboard shortcuts like Backspace, Ctrl+C etc.
- [x] Resizing columns triggers selection. Sometimes at the right-side, but always at top or bottom.
- [x] FEEL Input onBlur not working. Typing a value and clicking outside does not save the typed value.
- [x] Pressing `Esc` when editing an Annotations column name breaks the keyboard navigation.
- [x] When editing a Relation/DT cell, if you click outside it (blur) to set the value, it only works if the clicked cell is to the left or above the active cell.
- [x] Selection bug when adding/deleting rows/column.

##### UX:

- [x] Pasting expressions error handling. Must be done right at the menu. No modals/popovers.
- [x] HitPolicy selector should not close the Popover when selecting values on the Selects.
- [x] When double-clicking a column resize handler, resize it to fit to data.

##### UI:

- [x] Ctrl+ and Ctrl- zoom problems. Zoom in and out.
- [x] FEEL expressions highlight is wrong. E.g. "anything" -> "in" is blue, rest is black.
- [x] Add empty state to parameters popover on Function expressions.
- [x] Add icons to FunctionKind selector. (FEEL, Java, and PMML)
- [x] Better labels for HitPolicy selector. All of them should have a good short description too.
- [x] Better labels for FunctionKind selector. All of them should have a good short description too.
- [x] Add description to all expression types, right at the `Select expression` menu.

### On DMN Editor:

- [x] Keyboard shortcuts are not working.
- [x] Undo/Redo
- - [x] Re-enable Undo/Redo experience
- - [x] Doesn't work when resizing the Table
- - [x] TypeRef (and TypeRef holder) updates from react are not correctly managed in the case of Custom Types
- - [x] Undo/Redo doesn't work correctly when resetting an expression
- - [x] Autofilling root expressions based on Data Types
- - [x] Propagate expression updates in React only if the Expression actually changes.
- - [x] Undo/Redo for clear expression actions throws an exception
- - [x] DMN Elements IDs change during UNDO/REDO actions
- - [x] Undo/Redo doesn't work as expected in BKM nodes (It reproduces in the previous React BEE version as well)
- - [x] Redo a resizing doesn't work in some scenarios (e.g. Context expression with nested expressions)
- [x] Removing auto-filling ContextEntry items with Literal Expression with null value and comment
- [x] Basic Autofilling nested expressions based on Data Types (Same capabilities of current GWT-based BEE)
- [x] Check this -> https://github.com/kiegroup/kie-issues/issues/42
- [x] Marshaller fails with undefined nested expression in List
- [x] Synchronization between old and new editors.
- [x] Fix integration tests

### On DMN Runner:

- [x] Context Menu is appearing too far from the clicked place.
- [x] Keyboard shortcuts are not working.
- [x] Infinite loop happening when rendering the table.
- [x] Input cells are losing focus after input.
- [x] Deleting a line doesn't work.
- [x] `setRows(prev => {})` is not behaving well. DmnRunnerInputs should be able to provide the correct `prev` variable like setters of `useState` do. This impacts pasting/reseting cells, as only changes to the last impacted row are being persisted.
- [x] Column widths are not being persisted.

**Comment by tiagobento** [comment]
Just fixed Context Menus both on DMN Editor and DMN Runner. Found out that Keyboard Shortcuts work well if we stop using `useElementsThatStopKeyboardEventsPropagation`. Of course that implies pressing `D` while on the Boxed Expression Editor, for example, will create a Decision node, which is not ideal. Working on a fix for it now.

**Comment by tiagobento** [comment]
Removing `@patternfly/react-table` fixed the issue where the zoom is too large.

**Comment by yesamer** [comment]
Set the React-based Boxed Expression Component as the default choice: It is still possible to switch to the old one

https://user-images.githubusercontent.com/16005046/216011003-8cc0e95a-a5f3-4044-b67b-7495c43c22d2.mov

**Comment by yesamer** [comment]
**Undo-Redo Restored** 

https://user-images.githubusercontent.com/16005046/216094697-c678f55c-9eec-4b09-a8c4-fce19a563eb5.mov

I modified the involved API name (`notifyUserChange()` -> `createUndoCommand`) and I wrote a comment that better clarifies its usage. Feel free to modify them. 
Testing Undo Redo, I found a couple of minor bugs we need to address, I'll update the above list accordingly.

(apologies for the low-resolution video, I needed to save space :p )
